rrdtool/rrd_mts_similarity.c

```diff
@@ -0,0 +1,483 @@
+		   i, rrdms[i].mts_average, rrdms[i].mts_stddev,
+		   j, rrdms[j].mts_average, rrdms[j].mts_stddev);
+	  
+	num_potentially_zero_equal++;
+      } else if(circles_touch(rrdms[i].mts_average, rrdms[i].mts_stddev, rrdms[j].mts_average, rrdms[j].mts_stddev)
+		) {
+	float similarity = ndpi_bin_similarity(&rrdms[i].mts_b, &rrdms[j].mts_b, 0, similarity_threshold);
+
+	if((similarity >= 0) && (similarity < similarity_threshold)) {
+	  if(verbose)
+	    printf("Host: %d [%.1f/%.1f]  - Host: %d [%.1f/%.1f] are %s [%.1f]\n",
+		   i, rrdms[i].mts_average, rrdms[i].mts_stddev,
+		   j, rrdms[j].mts_average, rrdms[j].mts_stddev,
+		   (similarity == 0) ? "alike" : "similar",
+		   similarity
+		   );
+
+	  num_similar_mts++;
+	}
+      }
+    }
+  }
+ 
+  printf("Found %u (%.3f %%) similar Multivariates / %u zero alike Multivariates [num_mts: %d]\n",
+	 num_similar_mts,
+	 (num_similar_mts*100.)/(float)(num_host*num_host),
+	 num_potentially_zero_equal,
+	 num_host);
+}
+
+/* *************************************************** */
+
+int find_rrds(char *basedir, char *filename[], rrd_multifile_stats *rrdms, int *num_rrds, u_int *num_tot_rrds, int n_file, int *num_host) {
+  struct dirent **namelist;
+  int n = scandir(basedir, &namelist, 0, NULL);
+  u_int i;
+  bool t = false;
+  char path[PATH_MAX];
+
+  if(n < 0)
+    return 0; /* End of the tree */
+
+  while(n--) {
+    if(namelist[n]->d_name[0] != '.') {
+      struct stat s;
+
+      snprintf(path, sizeof(path), "%s/%s", basedir, namelist[n]->d_name);
+
+      if(stat(path, &s) == 0) {
+	if(S_ISDIR(s.st_mode))
+	  *num_host += find_rrds(path, filename, rrdms, num_rrds, num_tot_rrds, n_file, num_host);
+	else {
+	for(i=0; i<n_file; i++){
+	  if(strcmp(namelist[n]->d_name, filename[i]) == 0) {
+	    if(*num_tot_rrds < MAX_NUM_RRDS) {
+	      rrdms[*num_host].rfs[i].path = strdup(path);
+	      if(rrdms[*num_host].rfs[i].path != NULL)
+	        {
+	            t = true;
+	            (*num_tot_rrds)++;	//counter of all rrds
+	            (num_rrds[i])++;		//counter of each rrd
+	        }
+	      }
+	    }
+	  }
+	}
+      }
+    }
+    free(namelist[n]);
+  }
+  
+  free(namelist);
+
+  /* Report if I've found at least one of the files for a host */
+  if(t)
+    return 1;
+  else
+    return 0;
+}
+
+/* *************************************************** */
+
+int main(int argc, char *argv[]) {
+  rrd_time_value_t start_tv, end_tv;
+  char **filename, *start_s, *end_s, *dirname = NULL, *basedir = NULL;
+  u_int first = 1, quick_mode = 0;
+  float alpha;
+  char c;
+  time_t start, end;
+  u_int num_tot_rrds = 0, i, j = 0;
+  rrd_multifile_stats *rrdms;
+  int n_file = 0, num_host = 0, *num_rrds;
+
+  /* Defaults */
+  alpha   = DEFAULT_ALPHA;
+  start_s = DEFAULT_START;
+  end_s   = DEFAULT_END;
+
+  filename = malloc(sizeof(char*)*MAX_NUM_FILE);
+  for(i=0; i<MAX_NUM_FILE; i++)
+  {
+    filename[i] = malloc(sizeof(char)*50);
+    filename[i] = NULL;
```

Comment:
This code has no sense (first allocate then set it to zero). Consider using strdup()

Reply:
Fixed in commit with ID 030a0c3